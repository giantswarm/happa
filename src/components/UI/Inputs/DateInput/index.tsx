import {
  DateInput as Input,
  FormField,
  FormFieldProps,
  MaskedInputType,
  ThemeContext,
  ThemeType,
} from 'grommet';
import * as React from 'react';
import { css, DefaultTheme } from 'styled-components';

interface IMaskedInputProps {
  theme: DefaultTheme;
  focus?: boolean;
}

const customTheme: ThemeType = {
  global: {
    drop: {
      background: 'input-background',
      shadowSize: 'none',
      border: {
        radius: '0px 0px 4px 4px',
      },
      zIndex: '9999',
      // @ts-expect-error
      extend: css`
        border: 1px solid ${({ theme }) => theme.global.colors.border.dark};
        padding: 16px;
      `,
    },
  },
  dateInput: {
    icon: {
      size: 'small',
    },
  },
  maskedInput: {
    // @ts-expect-error
    extend: css`
      background: ${({ theme }) => theme.global.colors['input-background']};
      border-bottom-left-radius: ${(props: IMaskedInputProps) =>
        props.focus && 0};
      border-bottom-right-radius: ${(props: IMaskedInputProps) =>
        props.focus && 0};
    `,
  },
};

interface IBaseProps
  extends Omit<
    React.ComponentPropsWithoutRef<typeof Input>,
    'inputProps' | 'size' | 'onBlur'
  > {
  value: string | string[];
}

interface IDateInputProps
  extends IBaseProps,
    Omit<MaskedInputType, keyof IBaseProps> {
  /**
   * The description text displayed above the input.
   */
  label?: FormFieldProps['label'];
  /**
   * Props to be passed to the text input wrapper.
   */
  contentProps?: FormFieldProps['contentProps'];
  /**
   * Props to be passed to the form field.
   */
  formFieldProps?: FormFieldProps & { className?: string };
  /**
   * Whether the input is required or not.
   */
  required?: Exclude<FormFieldProps['required'], { indicator: boolean }>;
  /**
   * An error to be displayed below the input.
   */
  error?: FormFieldProps['error'];
  /**
   * A description to be displayed below the input.
   */
  info?: FormFieldProps['info'];
  /**
   * A help text to be displayed between the label
   * and the input.
   */
  help?: FormFieldProps['help'];
  /**
   * The margin between the form field and other elements.
   */
  margin?: FormFieldProps['margin'];
  /**
   * Whether the input should have some extra padding or not.
   */
  pad?: FormFieldProps['pad'];
  /**
   * How big should the input and its contents be.
   */
  size?: MaskedInputType['size'];
}

const DateInput = React.forwardRef<HTMLInputElement, IDateInputProps>(
  (
    {
      id,
      label,
      contentProps,
      formFieldProps,
      disabled,
      required,
      error,
      info,
      help,
      name,
      margin,
      pad,
      value,
      onChange,
      buttonProps,
      calendarProps,
      defaultValue,
      dropProps,
      format,
      inline,
      ...props
    },
    ref
  ) => {
    return (
      <ThemeContext.Extend value={customTheme}>
        <FormField
          htmlFor={id}
          label={label}
          contentProps={contentProps}
          disabled={disabled}
          required={required}
          name={name}
          error={error}
          info={info}
          help={help}
          margin={margin}
          pad={pad}
          {...formFieldProps}
        >
          <Input
            ref={ref}
            id={id}
            name={name}
            value={value}
            onChange={onChange}
            buttonProps={buttonProps}
            calendarProps={{
              animate: true,
              firstDayOfWeek: 1,
              daysOfWeek: true,
              ...calendarProps,
            }}
            defaultValue={defaultValue}
            format={format}
            inline={inline}
            inputProps={{
              ...props,
              disabled: disabled,
              required: required,
            }}
            dropProps={{
              align: { top: 'bottom' },
              ...dropProps,
            }}
          />
        </FormField>
      </ThemeContext.Extend>
    );
  }
);

DateInput.defaultProps = {
  size: 'medium',
};

export default DateInput;
