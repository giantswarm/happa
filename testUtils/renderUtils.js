import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
