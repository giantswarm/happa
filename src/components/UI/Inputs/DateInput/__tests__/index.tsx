import { renderWithTheme } from 'test/renderUtils';

import DateInput from '..';

describe('DateInput', () => {
  it('renders a simple input', () => {
    const { container } = renderWithTheme(DateInput, {
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a label', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      label: 'Date of birth',
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a validation error', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      label: 'Date of birth',
      error: 'You failed at this',
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with a help message', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      label: 'Date of birth',
      help: 'This is a very cool input',
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders an input with an info message', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      label: 'Date of birth',
      info: 'You can set a cool value',
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a disabled input', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      label: 'Date of birth',
      disabled: true,
      format: 'yyyy-mm-dd',
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders a button that triggers the calendar', () => {
    const { container } = renderWithTheme(DateInput, {
      id: 'dob',
      buttonProps: {
        label: 'Pick a date and call me Jim',
      },
      value: '',
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
