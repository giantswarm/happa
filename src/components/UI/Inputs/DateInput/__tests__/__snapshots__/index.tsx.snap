// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput renders a button that triggers the calendar 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 gCbCFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <button
          aria-label="Open Drop"
          class="StyledButtonKind-sc-1vhfpnt-0 caOOer"
          id="dob"
          type="button"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 jLKoGz"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 bgCrCI"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 fouGwC"
            />
            Pick a date and call me Jim
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput renders a disabled input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <label
        class="StyledText-sc-1sadyjn-0 icUwKR"
        for="dob"
      >
        Date of birth
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 higGFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 izCvfP"
              disabled=""
              id="dob"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput renders a simple input 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 gCbCFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 dUzPvu"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput renders an input with a help message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <label
        class="StyledText-sc-1sadyjn-0 icUwKR"
        for="dob"
      >
        Date of birth
      </label>
      <span
        class="StyledText-sc-1sadyjn-0 eASVNJ"
      >
        This is a very cool input
      </span>
      <div
        class="StyledBox-sc-13pk1d4-0 gCbCFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 dUzPvu"
              id="dob"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput renders an input with a label 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <label
        class="StyledText-sc-1sadyjn-0 icUwKR"
        for="dob"
      >
        Date of birth
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 gCbCFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 dUzPvu"
              id="dob"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput renders an input with a validation error 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <label
        class="StyledText-sc-1sadyjn-0 icUwKR"
        for="dob"
      >
        Date of birth
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 fuDwSj FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 dUzPvu"
              id="dob"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 fbPnYz"
      >
        You failed at this
      </span>
    </div>
  </div>
</div>
`;

exports[`DateInput renders an input with an info message 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 erdTQn"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 hYLvQc"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 ctdetp FormField__FormFieldBox-sc-m9hood-0 casIFC"
    >
      <label
        class="StyledText-sc-1sadyjn-0 icUwKR"
        for="dob"
      >
        Date of birth
      </label>
      <div
        class="StyledBox-sc-13pk1d4-0 gCbCFn FormField__FormFieldContentBox-sc-m9hood-1"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 dlgYyt"
        >
          <div
            class="StyledMaskedInput__StyledMaskedInputContainer-sc-99vkfa-1 dnGFZX"
          >
            <input
              autocomplete="off"
              class="StyledMaskedInput-sc-99vkfa-0 dUzPvu"
              id="dob"
              placeholder="yyyy-mm-dd"
              value=""
            />
          </div>
          <button
            class="StyledButtonKind-sc-1vhfpnt-0 gnGbFE"
            type="button"
          >
            <svg
              aria-label="Calendar"
              class="StyledIcon-sc-ofa7kd-0 hbMXBw"
              viewBox="0 0 24 24"
            >
              <path
                d="M2 5h20v17H2V5zm16 0V1M6 5V1m-4 9h20"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
      </div>
      <span
        class="StyledText-sc-1sadyjn-0 drMgMi"
      >
        You can set a cool value
      </span>
    </div>
  </div>
</div>
`;
